### ISSUE TITLE
workload__training

### ISSUE THREAD

**Issue by oyazdanb** [ROCm/aorta#238]
# Task — In-tree `training` Workload

## Goal

Add one public, in-tree workload named `training` that runs real PyTorch model training loops. The recipe chooses the distributed strategy:

```yaml
workload: training
workload_config:
  parallelism: fsdp   # ddp|fsdp
```

This workload is **not** `race`. The existing `race` workload remains the runtime corruption / overlap stress reproducer. `training` is for normal model training behavior: forward, loss, backward, optimizer, numeric checks, timings, and `WorkloadResult` reporting.

After this task:

- `aorta run --workload training --steps N` runs a real training loop.
- Recipes can choose `parallelism: ddp` or `parallelism: fsdp`.
- One-node smoke works without needing a multi-node cluster.
- Single-node multi-rank smoke works under `torchrun`.

## Why One Workload

DDP and FSDP are variants of the same experiment: "run a real training loop with this parallelism strategy." Keeping them under one workload avoids duplicated model/config/training code and makes recipe comparisons easier.

Use the recipe for experiment shape:

- `parallelism`
- model topology
- optimizer
- dtype
- batch size
- steps
- numeric checks

Use sidecars only for ad-hoc registries: extra mitigations, environments, or temporary local definitions. Do not use sidecars to choose DDP vs FSDP or model structure.

## Files to create / modify

Follow the existing flat-module convention: every in-tree public workload
(`workloads/race.py`, `workloads/llm_determinism.py`, `workloads/_subprocess.py`)
is a **single module**, not a subpackage. `llm_determinism.py` already bundles
its config dataclass, helper classes, and the workload in one file — mirror that.
Do **not** introduce a `workloads/training/` subpackage.

```
src/aorta/workloads/training.py                 # NEW — TrainingWorkload(Workload)
                                                #       + typed config dataclass(es)
                                                #       + shared train-step helpers
recipes/example-training-fsdp-smoke.yaml        # NEW — one-node/single-node smoke recipe
recipes/example-training-ddp-smoke.yaml         # NEW — one-node/single-node smoke recipe
tests/workloads/test_training.py                # NEW — config, lifecycle, result schema
```

Note: the existing `src/aorta/training/` package (`fsdp_trainer.py`) is a
standalone CLI benchmark, NOT a library and NOT the workloads namespace. The new
workload lives at `src/aorta/workloads/training.py` and is the entry-point target.

**What to reuse from `aorta.training.fsdp_trainer` (5 clean functions):**

- `init_distributed()` — `dist.init_process_group`, device assignment, rank/world_size resolution. Import directly.
- `build_fsdp_model()` / `build_ddp_model()` — copy the ~15-line FSDP/DDP wrapping block from each; do NOT import them directly because they hardcode `RankingTransformerModel` inside their body. Build `RepeatedBlockModel` in the workload first, then apply the wrapping lines.
- `configure_optimizer()` — AdamW setup. Import directly; use the AdamW branch only, ignore the Shampoo branch.
- `configure_scheduler()` — LambdaLR warmup/decay. Import directly.

**What NOT to reuse from `fsdp_trainer`:**

- `training_loop()` — requires `StreamProfiler` as a mandatory arg (not optional), writes its own `.jsonl` per rank (duplicates the dispatcher), and carries `inject_allreduce_copies` reproducer noise. Write a fresh ~40-line train step instead.
- `compute_loss()` — specific to ranking (BCE + importance weights). The new workload uses cross-entropy over synthetic logits.
- `MetricsLogger` — writes per-rank `.jsonl`; conflicts with the dispatcher's rank-0 JSON contract.
- `StreamProfiler` / `KernelTraceProfiler` — benchmark infra, not needed here.
- `collect_rocm_metrics` / `setup_signal_handlers` — benchmark extras.

**Other reuse:**

- `aorta.models.repeated_block` (`RepeatedBlockModel` + `BlockConfig`) — dense-transformer and MoE (`num_experts > 1` → top-1 router) topologies for `transformer` / `moe_transformer`.
- `aorta.data.synthetic_dataset` (`SyntheticDatasetConfig`, `create_dataloader`) — public synthetic input/target tensors.
- The `mlp` smoke topology is the only thing that may need a small local builder if `repeated_block` is too heavy for the smallest lifecycle smoke.

The train step itself is ~40 lines: zero_grad → forward → loss → backward → optimizer.step → NaN check → timing. Write it inline in the workload; do not import `training_loop`.

Modify `pyproject.toml`:

```toml
[project.entry-points."aorta.workloads"]
training = "aorta.workloads.training:TrainingWorkload"
```

## Contract details

- **Base contract**: import `Workload` and `WorkloadResult` from `aorta.workloads`.
- **Name declaration**: set `name: ClassVar[str] = "training"` to match the existing convention (`RaceWorkload.name`, `LlmDeterminismWorkload.name`).
- **Launch declaration**: declare `launch_mode = "distributed"` and `min_world_size = 1`.
- **One-node smoke**: support singleton distributed smoke when `WORLD_SIZE` is absent or `1`. This lets `aorta run --workload training --steps 2` test lifecycle and JSON schema. It is not a meaningful distributed-performance signal.
- **Multi-rank path**: support `torchrun --standalone --nproc_per_node=2 $(which aorta) run --workload training --steps 2`.
- **Distributed ownership**: the workload owns `torch.distributed.init_process_group(...)`; the platform does not initialize distributed.
- **Device behavior**: use CUDA/ROCm when available. CPU/Gloo is allowed only for unit tests or explicit smoke config; do not silently claim GPU coverage from CPU-only runs.
- **`--steps` flow**: `--steps S` from `aorta run` arrives as `self.config["steps"]` and overrides recipe/default step count.
- **Rank-0 output**: dispatcher writes trial JSON on rank 0 only. The workload should not write its own trial JSON.

## Recipe shape

```yaml
schema_version: 1
ticket: EXAMPLE-TRAINING-FSDP
workload: training
trials: 1
steps: 4

workload_config:
  parallelism: fsdp        # ddp|fsdp
  seed: 1234
  device: auto            # auto|cuda|cpu
  dtype: bfloat16         # bfloat16|float16|float32
  batch_size: 2
  seq_len: 128
  warmup_steps: 1

  model:
    kind: transformer     # mlp|transformer|moe_transformer
    hidden_size: 256
    num_layers: 2
    num_heads: 4
    ffn_size: 1024
    moe:
      enabled: false
      num_experts: 4
      # NOTE: aorta.models.repeated_block MoE is top-1 only — there is no
      # top_k knob. Drive MoE via num_experts; do not expose top_k unless
      # the underlying model gains top-k routing.

  optimizer:
    kind: adamw
    lr: 0.0001
    weight_decay: 0.01

  checks:
    fail_on_nan_loss: true
    fail_on_nan_grad: true
    fail_on_nonfinite_output: true

cells:
  - name: baseline-local
    mitigations: [none]
    environment: local
```

## Implementation guidance

Use typed config dataclasses for structured config. Avoid loose dict plumbing beyond the `Workload(config)` boundary.

Build topologies from existing in-tree models (`aorta.models.repeated_block`) and synthetic inputs (`aorta.data.synthetic_dataset`) — do not author fresh model code:

- `mlp`: smallest lifecycle smoke (small local builder acceptable if `repeated_block` is too heavy).
- `transformer`: attention + FFN path. Map onto `RepeatedBlockModel` with `num_experts = 1` (dense FFN).
- `moe_transformer`: map onto `RepeatedBlockModel` with `num_experts > 1` (top-1 router). The model is **top-1 only** — no `top_k`.

Keep DDP/FSDP branching narrow. `build_fsdp_model` / `build_ddp_model` in `fsdp_trainer.py` hardcode `RankingTransformerModel` inside their body — do not import them directly, as that would require modifying a standalone benchmark. Instead, build the model first in the workload, then copy just the ~15-line FSDP/DDP wrapping block from those functions:

- build `RepeatedBlockModel` from recipe config
- if `parallelism == "ddp"`, apply the DDP wrapping lines from `build_ddp_model`
- if `parallelism == "fsdp"`, apply the FSDP wrapping lines from `build_fsdp_model`
- share the same inline train step after wrapping

Write a fresh ~40-line train step inline (do NOT call `training_loop` from `fsdp_trainer` — it requires `StreamProfiler` as a mandatory arg and writes its own per-rank `.jsonl`):

```
for step in range(steps):
    t0 = time.perf_counter()
    optimizer.zero_grad(set_to_none=True)
    outputs = model(inputs)
    loss = F.cross_entropy(outputs.view(-1, vocab_size), targets.view(-1))
    loss.backward()
    optimizer.step()
    step_ms = (time.perf_counter() - t0) * 1000
    # NaN/inf checks on loss, outputs, gradients
    # collect step_ms, update executed_iterations
```

Import `configure_optimizer` / `configure_scheduler` / `init_distributed` from `aorta.training.fsdp_trainer` directly.

Do not copy `src/aorta/race/modes/fsdp.py`. It is a reproducer reference, not the implementation base for real training.

## Acceptance criteria

- [ ] `training` entry point exists in `pyproject.toml`.
- [ ] `TrainingWorkload(Workload)` implements `setup()`, `run()`, and `cleanup()`.
- [ ] `TrainingWorkload` declares `name = "training"`, `launch_mode = "distributed"`, and `min_world_size = 1`.
- [ ] Workload lives in a single flat module `src/aorta/workloads/training.py` (no subpackage).
- [ ] Model topologies reuse `aorta.models.repeated_block` / `aorta.data.synthetic_dataset`.
- [ ] DDP/FSDP wrapping copies the ~15-line wrapping block from `build_ddp_model` / `build_fsdp_model` in `fsdp_trainer.py`; those functions are NOT imported directly (they hardcode `RankingTransformerModel`).
- [ ] `configure_optimizer` / `configure_scheduler` / `init_distributed` imported from `aorta.training.fsdp_trainer`.
- [ ] Train step is written inline (~40 lines); `training_loop` from `fsdp_trainer` is NOT called.
- [ ] Recipe config supports `parallelism: ddp|fsdp`.
- [ ] Recipe config supports at least `model.kind: mlp|transformer`; `moe_transformer` can be included if scoped small.
- [ ] Bare one-rank smoke works: `aorta run --workload training --trials 1 --steps 2`.
- [ ] Single-node multi-rank smoke works for DDP and FSDP under `torchrun`.
- [ ] `run()` returns `WorkloadResult` with `passed`, `failure_count`, `failure_details`, `step_times_ms`, `total_iterations`, `elapsed_sec`, `main_work_started`, `executed_iterations`, and `configured_iterations`.
- [ ] Metrics include `parallelism`, `rank`, `world_size`, `device`, `dtype`, `model_kind`, `parameter_count`, `final_loss`, `step_time_p50`, and `step_time_p99`.
- [ ] Numeric checks fail on NaN/inf loss, NaN/inf gradients, or non-finite model outputs.
- [ ] `race`, `race mode=fsdp`, and `aorta run --workload race` keep working.
- [ ] Tests cover config parsing, model construction, DDP/FSDP selection, singleton smoke, mocked/stubbed multi-rank setup, and WorkloadResult schema.

## Out of scope

- Customer/NDA model structures or traces.
- Full TorchRec recommendation model.
- HSDP, tensor parallelism, pipeline parallelism, ZeRO, or hybrid parallelism.
- Throughput tuning.
- Replacing the existing `race` workload.
- `search_space()` / `objective()` methods.

## How to test

```bash
# Unit tests:
python -m pytest tests/workloads/test_training.py -v

# One-rank smoke:
aorta run --workload training --trials 1 --steps 2

# Single-node DDP smoke:
torchrun --standalone --nproc_per_node=2 $(which aorta) run \
  --workload training --trials 1 --steps 2

# Recipe dry runs:
aorta triage run --recipe recipes/example-training-ddp-smoke.yaml --dry-run
aorta triage run --recipe recipes/example-training-fsdp-smoke.yaml --dry-run

# Single-node recipe smoke:
torchrun --standalone --nproc_per_node=2 $(which aorta) triage run \
  --recipe recipes/example-training-fsdp-smoke.yaml

# Result shape:
python -m json.tool results/training/trial_d0_m0_t0.json
```

**Comment by vivekkhandelwal1** [comment]
Closed by https://github.com/ROCm/aorta/pull/247.
